### PR TITLE
expose h0 in dynamic_lstm

### LINKED ISSUE #11339: incorrect assert h0 in Python API dynamic_gru
h0.shape should be `[-1, size]`

https://github.com/PaddlePaddle/Paddle/blob/59e10922b41f141ae25b5266275f10921a30d92a/python/paddle/fluid/layers/nn.py#L680-L685

### LINKED ISSUE #11340: Expose the input h0 of dynamic_gru op in Python API
The input `h0` was already implemented in C++ code, need to expose it in Python API.

### PR DESCRIPTION
**Pull request by Yancey1989**
Fixed #11340 
Fixed #11339 
Related issue: #11335

### CODE REVIEW

File: python/paddle/fluid/layers/nn.py [hunk 1]
```diff
@@ -387,12 +395,19 @@ def dynamic_lstm(input,
     cell = helper.create_tmp_variable(dtype)
     batch_gate = helper.create_tmp_variable(dtype)
     batch_cell_pre_act = helper.create_tmp_variable(dtype)
+    inputs = {'Input': input, 'Weight': weight, 'Bias': bias}
+    if h_0:
+        assert h_0.shape == (-1, size), \
```
Comment (guoshengCS) [inline python/paddle/fluid/layers/nn.py]: The first dimension represents placeholder of batch size and may not be -1,  `input.shape[0]` might be better than `-1` thus we needn't assume users' batch size setting.

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -272,7 +272,9 @@ def dynamic_lstm(input,
                  cell_activation='tanh',
                  candidate_activation='tanh',
                  dtype='float32',
-                 name=None):
+                 name=None,
+                 h_0=None,
+                 c_0=None):
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: Put the name at the end?

Remove the TODO in line 264.

File: python/paddle/fluid/layers/nn.py (hunk 1 above)
Comment (qingqing01) [reply python/paddle/fluid/layers/nn.py]: Yeah, if the shape has been checked in the `InferShape`, maybe the Python doesn't need to check it.

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -358,6 +360,12 @@ def dynamic_lstm(input,
         dtype(str): Data type. Choices = ["float32", "float64"], default "float32".
         name(str|None): A name for this layer(optional). If set None, the layer
                         will be named automatically.
+        h_0(Variable): The initial hidden state is an optional
+                       input. This is a tensor with shape (N x D), where N is the
+                       batch size and D is the hidden size.
+        c_0(Variable): The initial cell state is an optional
+                       input. This is a tensor with shape (N x D), where N is the
+                       batch size. `h_0` and `c_0` can be NULL but only at the same time.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: Tell the users if not set, the default is zero.

File: python/paddle/fluid/layers/nn.py (hunk 1 above)
Comment (Yancey1989) [reply python/paddle/fluid/layers/nn.py]: `Infershape` doesn't check the shape, we can check it in Python code, it's more clear for the user.

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (Yancey1989) [reply python/paddle/fluid/layers/nn.py]: Done.

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (Yancey1989) [reply python/paddle/fluid/layers/nn.py]: Done.

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -676,10 +692,12 @@ def dynamic_gru(input,
...
     bias = helper.create_parameter(
         attr=helper.bias_attr, shape=[1, 3 * size], dtype=dtype, is_bias=True)
+    batch_size = input.shape[0]
     inputs = {'Input': input, 'Weight': weight, 'Bias': bias}
     if h_0 != None:
         assert h_0.shape == (
-            size, size), 'The shape of h0 should be(%d, %d)' % (size, size)
+            batch_size, size
+        ), 'The shape of h0 should be(batch_size, %d)' % size
         inputs['h0'] = h_0
```
Comment (guoshengCS) [inline python/paddle/fluid/layers/nn.py]: It also should be `inputs['H0'] = h_0` like in LSTM. Sorry for leaving out this.

Comment (Yancey1989) [reply python/paddle/fluid/layers/nn.py]: Thanks! Done.

**Pull request merged**